### PR TITLE
C4 nourhan z c4 f2 record vote

### PR DESCRIPTION
**Pull request by nourhanZ**
#96

### CODE REVIEW

File: arability/app/models/vote.rb [hunk 1]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+
+   # Assigned to: Nourhan Zakria
+   # This method is used to record the vote given for a certain synonym by a ceratin user
+   # Parameters:
+   #  gamer_id: the voter(gamer) ID
+   #  synonym_id: the synonym_id that the gamer voted for
+   # Returns:
+   #  On success: Returns true and the instance of vote that was created and saved
+   #  On failure: returns false and the instance of vote that wasn't saved
+  def self.record_vote(gamer_id, synonym_id)
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: use class << self rather than self.

File: arability/app/models/vote.rb [hunk 2]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+   #  On success: Returns true and the instance of vote that was created and saved
+   #  On failure: returns false and the instance of vote that wasn't saved
+  def self.record_vote(gamer_id, synonym_id) 
+          @vote = Vote.new
+          @vote.synonym_id = synonym_id
+          @vote.gamer_id = gamer_id
+            if @vote.save
+              return true, @vote  
+            else
+              return false, @vote
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: properly indent the "end" keywords

File: arability/app/models/vote.rb [hunk 3]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+   # This method is used to retreive a list of specific size of keywords that gamer with this gamer_id didn't vote on yet
+   # Parameters:
+   #  gamer_id: the gamer ID 
+   #  count: the size of the list to be retreived
+   #  lang: integer to specify the language of keywords to be retreived if 0 then english only, if 1 then arabic only, otherwise both english and arabic 
+   #        keywords can be icluded
+   # Returns:
+   #  On success: Returns a list of un voted keywords of the specified langauge with size = count for the gamer with this gamer_id  
+   #  On failure: Returns empty list
+  def self.get_unvoted_keywords(gamer_id, count, lang)
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: something like list_size would be more descriptive than just count

File: arability/app/models/vote.rb [hunk 4]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+      if Gamer.find_by_id(gamer_id) != nil
+        voted_synonyms = Vote.where('gamer_id = ?', gamer_id).select('synonym_id')
+        voted_keywords = Synonym.where(:id => voted_synonyms).select('keyword_id')
+        #un_voted_keywords = Keyword.all - Keyword.where( :id => voted_keywords)
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: remove commented line

File: arability/app/models/vote.rb [hunk 5]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+          if lang == 0
+            un_voted_keywords = Keyword.where("is_english =?",true) - Keyword.where(:id => voted_keywords)
+          elsif lang ==1
+            un_voted_keywords = Keyword.where("is_english=?",false) - Keyword.where(:id => voted_keywords)
+          else
+            un_voted_keywords = Keyword.all - Keyword.where( :id => voted_keywords) 
+          end
+        list_returned = un_voted_keywords.take(count)
+
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb:13]: remove extra blank lines

File: arability/app/models/vote.rb [hunk 6]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+   # Returns:
+   #  On success: Returns a list of un voted keywords of the specified langauge with size = count for the gamer with this gamer_id  
+   #  On failure: Returns empty list
+  def self.get_unvoted_keywords(gamer_id, count, lang)
+      if Gamer.find_by_id(gamer_id) != nil
+        voted_synonyms = Vote.where('gamer_id = ?', gamer_id).select('synonym_id')
+        voted_keywords = Synonym.where(:id => voted_synonyms).select('keyword_id')
+        #un_voted_keywords = Keyword.all - Keyword.where( :id => voted_keywords)
+          if lang == 0
+            un_voted_keywords = Keyword.where("is_english =?",true) - Keyword.where(:id => voted_keywords)
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: always leave a space after the commas

File: arability/app/models/vote.rb [hunk 7]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+      errors.add(:gamer_id,"this gamer_id soesn't exist")
+    end
+  end
+
+  # Written by: Nourhan Zakaria
+  #This is a custom validation method that validates that there exists a synonym with this synonym_id 
+  def validate_synonym_exists
+    valid_synonym = Synonym.find_by_id(synonym_id)
+    if valid_synonym == nil
+      errors.add(:synonym_id,"this synonym_id soesn't exist")
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: doesn't not soesn't :D correct the typo

File: arability/app/models/vote.rb [hunk 8]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+    end
+  end
+
+  # Written by: Nourhan Zakaria
+  # This is a custom validation method that validates that synonym which the gamer is voting for doesn't belong to a keyword that 
+  # this gamer voted for before
+  def validate_voting_for_new_keyword
+    keyword_id_of_chosen_synonym=Synonym.where("id=?",synonym_id).select('Keyword_id')
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: always use a space before and after the = sign

File: arability/app/models/vote.rb [hunk 9]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
   belongs_to :synonym
   belongs_to :gamer
   # attr_accessible :title, :body
+  validate :validate_gamer_exists, :validate_synonym_exists, :validate_voting_for_new_keyword
+  validates :gamer_id, :uniqueness => { :scope => :synonym_id, :message => "This gamer has aleready voted for this sysnonym before"} 
+
+   # Assigned to: Nourhan Zakria
```
Comment (mostafahassaan) [inline arability/app/models/vote.rb]: change "Assigned to" to Author

File: arability/app/models/vote.rb [hunk 10]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+          @vote.synonym_id = synonym_id
+          @vote.gamer_id = gamer_id
+            if @vote.save
+              return true, @vote  
+            else
+              return false, @vote
+   	        end
+  end
+
+   # This method is used to retreive a list of specific size of keywords that gamer with this gamer_id didn't vote on yet
```
Comment (mostafahassaan) [inline arability/app/models/vote.rb]: author

File: arability/app/models/vote.rb [hunk 11]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+        list_returned = un_voted_keywords.take(count)
+
+        return list_returned
+
+      else
+        return nil
+      end
+  end
+
+  # Written by: Nourhan Zakaria
```
Comment (mostafahassaan) [inline arability/app/models/vote.rb]: Change to Author! :D

File: arability/app/models/vote.rb [hunk 12]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+  #This is a custom validation method that validates that there exists a gamer with this gamer_id
```
Comment (mostafahassaan) [inline arability/app/models/vote.rb]: Add params, succes and fail. If It doens't have any params or a fail scenario add a "--" beside it

File: arability/app/models/vote.rb [hunk 13]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+  # Written by: Nourhan Zakaria
+  #This is a custom validation method that validates that there exists a gamer with this gamer_id 
+  def validate_gamer_exists
+    valid_gamer = Gamer.find_by_id(gamer_id)
+    if valid_gamer == nil
+      errors.add(:gamer_id,"this gamer_id soesn't exist")
+    end
+  end
+
+  # Written by: Nourhan Zakaria
```
Comment (mostafahassaan) [inline arability/app/models/vote.rb]: authod
Add params, succes and fail. If It doens't have any params or a fail scenario add a "--" beside it

File: arability/app/models/vote.rb [hunk 14]
```diff
@@ -2,4 +2,87 @@ class Vote < ActiveRecord::Base
...
+  #This is a custom validation method that validates that there exists a synonym with this synonym_id 
+  def validate_synonym_exists
+    valid_synonym = Synonym.find_by_id(synonym_id)
+    if valid_synonym == nil
+      errors.add(:synonym_id,"this synonym_id soesn't exist")
+    end
+  end
+
+  # Written by: Nourhan Zakaria
```
Comment (mostafahassaan) [inline arability/app/models/vote.rb]: authod
Add params, succes and fail. If It doens't have any params or a fail scenario add a "--" beside it

File: arability/app/models/vote.rb [hunk 15]
```diff
@@ -2,4 +2,77 @@ class Vote < ActiveRecord::Base
...
+    #  On failure: returns false and the instance of vote that wasn't saved
+    def record_vote(gamer_id, synonym_id) 
+      vote = Vote.new
+      vote.synonym_id = synonym_id
+      vote.gamer_id = gamer_id
+        if vote.save
+          return true, vote  
+        else
+          return false, vote
+ 	      end
```
Comment (m0hamed) [inline arability/app/models/vote.rb]: extra space here

File: arability/app/models/vote.rb [hunk 16]
```diff
@@ -2,4 +2,77 @@ class Vote < ActiveRecord::Base
...
+      if chosen_synonym != []
+        k_id_of_chosen_synonym = chosen_synonym.first.keyword_id
+        check = Keyword.where("id NOT IN( 
+          SELECT K.id FROM Keywords K INNER JOIN Synonyms S ON 
+          S.keyword_id = K.id AND S.id IN(SELECT S1.id FROM synonyms S1 
+          INNER JOIN votes V ON S1.id = V.synonym_id INNER JOIN 
+          gamers G ON G.id = V.gamer_id AND G.id = ?)) AND id = ?",
+          gamer_id, k_id_of_chosen_synonym)
+        if check == []
+          errors.add(:keyword_id,"this user voted for this keyword before")
```
Comment (m0hamed) [inline arability/app/models/vote.rb]: gamer instead of user

File: arability/app/models/vote.rb [hunk 17]
```diff
@@ -2,4 +2,77 @@ class Vote < ActiveRecord::Base
...
+      chosen_synonym = Synonym.where("id=?",synonym_id)
+      if chosen_synonym != []
+        k_id_of_chosen_synonym = chosen_synonym.first.keyword_id
+        check = Keyword.where("id NOT IN( 
+          SELECT K.id FROM Keywords K INNER JOIN Synonyms S ON 
+          S.keyword_id = K.id AND S.id IN(SELECT S1.id FROM synonyms S1 
+          INNER JOIN votes V ON S1.id = V.synonym_id INNER JOIN 
+          gamers G ON G.id = V.gamer_id AND G.id = ?)) AND id = ?",
+          gamer_id, k_id_of_chosen_synonym)
+        if check == []
```
Comment (m0hamed) [inline arability/app/models/vote.rb]: call .exists? on the query object instead

File: arability/app/models/vote.rb [hunk 18]
```diff
@@ -2,4 +2,77 @@ class Vote < ActiveRecord::Base
   belongs_to :synonym
   belongs_to :gamer
   # attr_accessible :title, :body
+  validate :validate_gamer_exists, :validate_synonym_exists, 
+  :validate_voting_for_new_keyword
+  validates :gamer_id, :uniqueness => { :scope => :synonym_id, 
+  :message => "This gamer has aleready voted for this synonym before" }
```
Comment (m0hamed) [inline arability/app/models/vote.rb]: indent this line

File: arability/app/models/vote.rb [hunk 19]
```diff
@@ -2,4 +2,77 @@ class Vote < ActiveRecord::Base
   belongs_to :synonym
   belongs_to :gamer
   # attr_accessible :title, :body
+  validate :validate_gamer_exists, :validate_synonym_exists, 
+  :validate_voting_for_new_keyword
```
Comment (m0hamed) [inline arability/app/models/vote.rb]: indent this line

File: arability/app/models/vote.rb [hunk 20]
```diff
@@ -2,4 +2,77 @@ class Vote < ActiveRecord::Base
...
+    #  synonym_id: the synonym ID that the gamer voted for
+    #Returns:
+    #  On success: returns true and the instance of vote that 
+    #  was created and saved
+    #  On failure: returns false and the instance of vote that wasn't saved
+    def record_vote(gamer_id, synonym_id) 
+      vote = Vote.new
+      vote.synonym_id = synonym_id
+      vote.gamer_id = gamer_id
+        if vote.save
```
Comment (NourhanMohamed) [inline arability/app/models/vote.rb]: why is this indented?

File: arability/app/models/keyword.rb [hunk 21]
```diff
@@ -1,4 +1,5 @@
 class Keyword < ActiveRecord::Base
+  has_many :synonyms
```
Comment (m0hamed) [inline arability/app/models/keyword.rb:2]: has many synonyms is repeated twice

File: arability/app/models/vote.rb [hunk 22]
```diff
@@ -5,9 +5,35 @@ class Vote < ActiveRecord::Base
   @@ENG = 0
   @@ARABIC = 1
   @@BOTH = 2
-  
+
+  validate :validate_gamer_exists, :validate_synonym_exists, 
+    :validate_voting_for_new_keyword
+  validates :gamer_id, :uniqueness => { :scope => :synonym_id,
```
Comment (m0hamed) [inline arability/app/models/vote.rb:7]: please put the validate statements above the constants

**Pull request merged**